### PR TITLE
Close #6518: autodoc: Add autodoc_type_aliases

### LINKED ISSUE #6518: Option for not unfolding aliases
Would it be possible to add an option for autodoc not to unfold user-defined type aliases? 
For example, if I introduce a type synonym Position = int and then define a method with argument pos: Position then I would like to see this typing in the documentation and not pos: int. For me, unfolding the alias is loosing information on how the program is built, something a documentation should not do, unless required by the author.

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Feature

### Purpose
- refs: #6518 
- autodoc_type_aliases allows to keep user defined type alises not
    evaluated in the generated document.

### CODE REVIEW

File: tests/test_ext_autodoc_configs.py [hunk 1]
```diff
@@ -642,6 +642,42 @@ def test_autodoc_typehints_description_for_invalid_node(app):
...
+        '   :module: target.annotations',
+        '',
+        '   docstring',
+        '',
+    ]
+
+
+@pytest.mark.sphinx('text', testroot='ext-autodoc',
+                    confoverrides={'autodoc_typehints_evaluation': False})
+def test_autodoc_typehints_evaluation_disabled(app):
```
Comment (tk0miya) [inline tests/test_ext_autodoc_configs.py]: This case should be invoked only with py3.7 or above.

File: doc/usage/extensions/autodoc.rst [hunk 2]
```diff
@@ -502,6 +502,44 @@ There are also config values that you can set:
...
+   When disabled, it generates a following document internally::
+
+     .. py:function:: f() -> AliasType:
+
+        ...
+
+   .. note::
+
+      As a side effect of disabling, cross references from function signature would
+      not work correctly.
```
Comment (tk0miya) [inline doc/usage/extensions/autodoc.rst]: We need to describe this feature more carefully.

**Comment by sjjessop** [comment]
Like I said over on the ticket, personally for me disabling all forward reference evaluation is overkill (especially in combination with `from __future__ import annotations`, which means that *all* annotations are forward references). I haven't tried the code in this branch, but in my own experiments with patching `get_type_hints`, suppressing all reference expansion prevents the annotation in the documented signature from linking to the thing referenced (which is what you already said about cross references, I think). This isn't too bad for an alias (especially if the reader can find it anyway in the same file), but not so good to have no cross-links to classes any more.

So ideally, I'd like some means of marking up a particular symbol (or if that's not possible, specify the symbol in the Sphinx config) to say, "I don't want this expanded, thanks". I'm afraid I don't have any suggestions how to implement that, though.

Even better IMO would be for `typing` to provide non-transparent type aliases, so that I can say `MyAlias = typing.Alias(int)` to mean, "MyAlias behaves in all respects like int for type-checking purposes, but its `__name__` is different". But that would presumably create work for every type-checking tool in the Python ecosystem, so I don't fancy my chances...

**Comment by tk0miya** [comment]
Thank you for your comment. I agree that this approach is overkill to realize type aliases. And I also understand about side-effects (cross references will be broken). They are why I still keep this draft version. I feel this option is not useful...

About "non-transparent type aliases": Recently, [PEP-613](https://www.python.org/dev/peps/pep-0613/) was accepted. It allows marking such type aliases via a special type: `TypeAlias`. So we can determine whether the variable is a type alias or not from annotations in the future.

**Comment by tk0miya** [comment]
I got a hint from @sjjessop 's comment https://github.com/sphinx-doc/sphinx/issues/6518#issuecomment-669861553 and I changed the approach now. The new PR adds a new configuration `autodoc_type_aliases` for defining type aliases. It keeps them not evaluated by the python interpreter and generates documents having as-is annotations.

**Pull request merged**